junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/ClassTestDescriptor.java

```diff
@@ -201,11 +203,73 @@ private Object instantiateAndPostProcessTestInstance(JupiterEngineExecutionConte
 		return instance;
 	}
 
+	protected Object invokeTestInstanceConstructor(Object outerInstance, ExtensionRegistry registry,
+			ExtensionContext extensionContext) {
+		Constructor<?> constructor = ReflectionUtils.getDeclaredConstructor(this.testClass);
+		if (outerInstance != null) {
+			return executableInvoker.invoke(constructor, outerInstance, extensionContext, registry);
+		}
+		else {
+			return executableInvoker.invoke(constructor, extensionContext, registry);
+		}
+	}
+
+	protected TestInstanceFactory resolveTestInstanceFactory(ExtensionRegistry registry,
+			ExtensionRegistry parentRegistry) {
+
+		List<TestInstanceFactory> factories = registry.getExtensions(TestInstanceFactory.class);
+		if (factories.isEmpty()) {
+			return null;
+		}
+
+		if (parentRegistry != null) {
+			List<TestInstanceFactory> parentFactories = parentRegistry.getExtensions(TestInstanceFactory.class);
+			factories.removeAll(parentFactories);
+		}
+
+		if (factories.size() > 1) {
+			String factoryNames = factories.stream().map(factory -> factory.getClass().getSimpleName()).collect(
+				Collectors.joining(","));
+
+			String errorMessage = String.format(
+				"Too many TestInstanceFactory extensions [%s] registered on test class: %s", factoryNames,
+				testClass.getSimpleName());
+
+			throw new JUnitException(errorMessage);
+		}
+
+		return factories.get(0);
+	}
+
+	protected Object invokeTestInstanceFactory(Object outerInstance, ExtensionRegistry registry,
+			ExtensionRegistry parentRegistry, ExtensionContext extensionContext) {
+
+		TestInstanceFactory factory = resolveTestInstanceFactory(registry, parentRegistry);
+		if (factory == null) {
+			return invokeTestInstanceConstructor(outerInstance, registry, extensionContext);
+		}
+		else {
+			try {
+				if (outerInstance != null) {
+					return factory.instantiateNestedTestClass(this.testClass, outerInstance, extensionContext);
+				}
+				else {
+					return factory.instantiateTestClass(this.testClass, extensionContext);
+				}
+			}
+			catch (RuntimeException exception) {
```

Comment:
Both `catch` are too generic. And please use `BlacklistedExceptions.rethrowIfBlacklisted(cause);`